bolt/lib/Passes/IdenticalCodeFolding.cpp

```diff
@@ -341,6 +341,61 @@ typedef std::unordered_map<BinaryFunction *, std::vector<BinaryFunction *>,
 namespace llvm {
 namespace bolt {
 
+Error IdenticalCodeFolding::processDataRelocations(
+    BinaryContext &BC, const SectionRef &SecRefRelData) {
+  for (const RelocationRef &Rel : SecRefRelData.relocations()) {
+    symbol_iterator SymbolIter = Rel.getSymbol();
+    const ObjectFile *OwningObj = Rel.getObject();
+    assert(SymbolIter != OwningObj->symbol_end() &&
+           "relocation Symbol expected");
+    const SymbolRef &Symbol = *SymbolIter;
+    const uint64_t SymbolAddress = cantFail(Symbol.getAddress());
+    const ELFObjectFileBase *ELFObj = dyn_cast<ELFObjectFileBase>(OwningObj);
+    if (!ELFObj)
+      return createFatalBOLTError(
+          Twine("BOLT-ERROR: Only ELFObjectFileBase is supported"));
+    const int64_t Addend = getRelocationAddend(ELFObj, Rel);
+    BinaryFunction *BF = BC.getBinaryFunctionAtAddress(SymbolAddress + Addend);
+    if (!BF)
+      continue;
+    BF->setUnsafeICF();
+  }
+  return Error::success();
+}
+
+Error IdenticalCodeFolding::markFunctionsUnsafeToFold(BinaryContext &BC) {
+  ErrorOr<BinarySection &> SecRelData = BC.getUniqueSectionByName(".rela.data");
```

Comment:
I don't understand why the processing is limited to `.rela.data`.

Reply:
As we discussed offline with static global/local function pointer gcc/clang can generate the references in the .rodata, but even with -O0 there are direct function references in the code.
Scanning .rela.rodata I think will add things like class destructors to skip list even thought they are safe to ICF.
Should the user have an option to disable check of this section?

Reply:
Are you saying it's impossible to have and application that loads a function address from read-only section without having the same function address being referenced directly from code?

Orthogonally, with pragma directives and linker scripts it's possible to assign arbitrary names to sections. Relying on section name matching in this case is wrong. If you want to check for read-only vs writable section, you should check section attributes.

Reply:
I mean nothing is impossible with all the stuff that ends up in prod binaries between various compilers and languages. It's just from some common patterns with c++ this was the case.
Changing code to process all the data sections, skipping ".plt", ".got.plt", ".eh_frame", ".gcc_except_table", ".data.rel.ro", ".fini_array", ".init_array"
Processing ".fini_array", ".init_array" seperatly to mark things in them as safe.
and running bolt on HHVM some stats:

BOLT-DEBUG: Relocation type: R_X86_64_64
BOLT-DEBUG: Relocation type: R_X86_64_PC32
BOLT-DEBUG: Section with PC: .rela.rodata
BOLT-DEBUG: Section with function references: .rela.data
BOLT-DEBUG: Section with function references: .rela.fini_array
BOLT-DEBUG: Section with function references: .rela.init_array
BOLT-DEBUG: Section with function references: .rela.rodata
BOLT-DEBUG: Section with function references: .rela.tdata

not marking function symbols from ".fini_array", ".init_array" as safe did not change size of what got folded.

Trying to figure out how PC32 gets generated in the .rela.rodata.

Reply:
> Trying to figure out how PC32 gets generated in the .rela.rodata.

Possibly from jump tables or indirect gotos.